System/AbstractFilePath/Data/ByteString/Short/Decode.hs

```diff
@@ -0,0 +1,286 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE BangPatterns #-}
+
+module System.AbstractFilePath.Data.ByteString.Short.Decode (decodeUtf16LE, decodeUtf16LEWith, decodeUtf16LE', decodeUtf16LE'', decodeUtf8, decodeUtf8With, decodeUtf8', strictDecode, lenientDecode, OnError, OnDecodeError, UnicodeException(..)) where
+
+import Data.ByteString.Internal
+    ( ByteString )
+import Data.ByteString.Short
+    ( ShortByteString, toShort )
+import Data.Bits
+    ( unsafeShiftL )
+import Data.Typeable (Typeable)
+import Data.Word
+    ( Word16, Word8 )
+import GHC.Word (Word8(..), Word16(..))
+import Control.Exception
+import qualified Data.ByteString.Short as BS
+    ( index, length )
+import GHC.IO
+    ( unsafeDupablePerformIO )
+import GHC.Base
+import Control.DeepSeq (NFData, rnf)
+import Numeric (showHex)
+#if !MIN_VERSION_base(4,16,0)
+
+word8ToWord#  :: Word# -> Word#
+word16ToWord# :: Word# -> Word#
+word8ToWord#  w = w
+word16ToWord# w = w
+{-# INLINE word16ToWord# #-}
+{-# INLINE word8ToWord# #-}
+
+#endif
+
+unsafeChr16 :: Word16 -> Char
+unsafeChr16 (W16# w#) = C# (chr# (word2Int# (word16ToWord# w#)))
+{-# INLINE unsafeChr16 #-}
+
+unsafeChr8 :: Word8 -> Char
+unsafeChr8 (W8# w#) = C# (chr# (word2Int# (word8ToWord# w#)))
+{-# INLINE unsafeChr8 #-}
+
+-- | /O(n)/ Convert a 'ByteString' into a 'Stream Char', using UTF-8
+-- encoding.
+streamUtf8 :: OnDecodeError -> ShortByteString -> [Char]
+streamUtf8 onErr bs = go 0
+    where
+      l = BS.length bs
+      go i
+          | i >= l = []
+          | validate1_8 x1 = (unsafeChr8 x1) : go (i+1)
+          | i+1 < l && validate2_8 x1 x2 = (chr2 x1 x2) : go (i+2)
+          | i+2 < l && validate3_8 x1 x2 x3 = (chr3 x1 x2 x3) : go (i+3)
+          | i+3 < l && validate4_8 x1 x2 x3 x4 = (chr4 x1 x2 x3 x4) : go (i+4)
+          | otherwise = decodeError "streamUtf8" "UTF-8" onErr (Just x1) ++ go (i+1)
+          where
+            x1 = idx i
+            x2 = idx (i + 1)
+            x3 = idx (i + 2)
+            x4 = idx (i + 3)
+            idx = BS.index bs
+{-# INLINE [0] streamUtf8 #-}
+
+-- | /O(n)/ Convert a 'ShortByteString' into a 'Stream Char', using little
+-- endian UTF-16 encoding.
+streamUtf16LE :: OnDecodeError -> ShortByteString -> [Char]
+streamUtf16LE onErr bs = go 0
+    where
+      l = BS.length bs
+      {-# INLINE go #-}
+      go i
+          | i >= l                         = []
+          | i+1 < l && validate1_16 x1    = (unsafeChr16 x1) : go (i+2)
+          | i+3 < l && validate2_16 x1 x2 = (chr2_16 x1 x2) : go (i+4)
+          | otherwise = decodeError "streamUtf16LE" "UTF-16LE" onErr Nothing ++ go (i+1)
+          where
+            x1    = idx i       + (idx (i + 1) `unsafeShiftL` 8)
+            x2    = idx (i + 2) + (idx (i + 3) `unsafeShiftL` 8)
```

Comment:
Not much point to use `unsafeShiftL` with a literal argument. A normal `shiftL` will produce the same Core.